### ISSUE TITLE
`-Wsometimes-uninitialized` suggests unhelpful removal of if statement

### ISSUE THREAD

**Issue by tbaederr** [llvm/llvm-project#62923]
See: https://godbolt.org/z/cGe68aK11

The output for:
```c++
int f1(int cond) {
  int a;
  if (cond) {


  } else {
    a = 3;
  }
  return a;
}
```

with `-Wsometimes-uninitialized` is:
```console
<source>:3:7: warning: variable 'a' is used uninitialized whenever 'if' condition is true [-Wsometimes-uninitialized]
  if (cond) {
      ^~~~
<source>:9:10: note: uninitialized use occurs here
  return a;
         ^
<source>:3:3: note: remove the 'if' if its condition is always false
  if (cond) {
  ^~~~~~~~~~~
<source>:2:8: note: initialize the variable 'a' to silence this warning
  int a;
       ^
        = 0
1 warning generated.
ASM generation compiler returned: 0
<source>:3:7: warning: variable 'a' is used uninitialized whenever 'if' condition is true [-Wsometimes-uninitialized]
  if (cond) {
      ^~~~
<source>:9:10: note: uninitialized use occurs here
  return a;
         ^
<source>:3:3: note: remove the 'if' if its condition is always false
  if (cond) {
  ^~~~~~~~~~~
<source>:2:8: note: initialize the variable 'a' to silence this warning
  int a;
       ^
        = 0
1 warning generated.
```

the suggestion to remove the if statement is is pretty useless, and just adds noise to the output. I think it would be pretty safe to simply remove this diagnostic.

/CC @cjdb @AaronBallman  @nikic

**Comment by AaronBallman** [comment]
Why do you think it's useless? There are two ways to correct the issue and either one could be "the right way to solve it" depending on context.

**Comment by tbaederr** [comment]
Oh sure, I understand why the diagnostic exists, it's just a pretty big leap to assume that the condition _might_ _maybe_ be always false and to suggest to outright remove the if statement. Assuming that the initialization was forgotten (the following note) makes much more sense.

(Side note: If the condition is actually always false, nothing is even emitted: https://godbolt.org/z/PcaKvbfhM. Adding a `std::unreachable()` in the if block makes the diagnostic go away, as expected)

**Comment by AaronBallman** [comment]
> Oh sure, I understand why the diagnostic exists, it's just a pretty big leap to assume that the condition _might_ _maybe_ be always false and to suggest to outright remove the if statement. Assuming that the initialization was forgotten (the following note) makes much more sense.

I think the second note is the more likely correct approach of the two, but I've seen just enough code like what you posted above where the correct answer is "remove the if" to think the note is defensible, but I also have no compelling data to back that up.

> (Side note: If the condition is actually always false, nothing is even emitted: https://godbolt.org/z/PcaKvbfhM. Adding a std::unreachable() in the if block makes the diagnostic go away, as expected)

Well that just makes me sad.

**Comment by tbaederr** [comment]
> > Oh sure, I understand why the diagnostic exists, it's just a pretty big leap to assume that the condition _might_ _maybe_ be always false and to suggest to outright remove the if statement. Assuming that the initialization was forgotten (the following note) makes much more sense.
> 
> I think the second note is the more likely correct approach of the two, but I've seen just enough code like what you posted above where the correct answer is "remove the if" to think the note is defensible, but I also have no compelling data to back that up.

I'm not gonna fight hard about this (I've never seen this diagnostic in my own development), but it seems like something that should be pointed out by a static analyzer that can actually proof that the condition is always false. If we can't prove it, suggesting the initialization is the right approach IMO.

But maybe @cjdb has another opinion.

**Comment by AaronBallman** [comment]
> I'm not gonna fight hard about this (I've never seen this diagnostic in my own development), but it seems like something that should be pointed out by a static analyzer that can actually proof that the condition is always false. If we can't prove it, suggesting the initialization is the right approach IMO.

I'd recommend looking at the review which added the functionality to see what discussions there were around the note (if any). FWIW, this is an analysis-based diagnostic, so it already uses control flow to determine whether it's initialized or not. However, it's not a static analysis diagnostic that does interprocedural analysis. I suspect the static analyzer could do a better job in cases that don't involve function parameters (or shared state like global variables).

**Comment by gustedt** [comment]
Just as another data point, I stumbled into a similar diagnostic today. 

First, this diagnostic is really not helpful for me, because the code is generated and it would be tedious to detect that in the generator.  So I can't remove the `if` branch easily.
Second,  in my situation the `if` branch is indeed never taken. I have the impression that gcc perfectly detects this (I am using  -fanalyzer, there) and that it removes the dead code without a diagnostic. (I suspect that clang also detects it, but much later, when the diagnostic is already launched.)
Third, the variable is actually uninitialized for a reason, because we want the analyzer to detect a case where indeed  the variable would not be initialized at some point.

**Comment by gustedt** [comment]
> I'm confused then -- that's how this diagnostic already works. If there is an uninitialized variable that remains uninitialized on a code path through the function, then a use of that variable will trigger the diagnostic.

The difference in expectation is that I want it to diagnose the use of an uninitialized variable in *reachable* code.

I see that it is not always possible to tell which part of the code is reachable or not, but I would appreciate if some effort in flow analysis could be made before there is a diagnosis.

(Yes, this is dumping the insufficiency of some code generation tools on you guys ;-)

Thanks
Jens

**Comment by AaronBallman** [comment]
> > I'm confused then -- that's how this diagnostic already works. If there is an uninitialized variable that remains uninitialized on a code path through the function, then a use of that variable will trigger the diagnostic.
> 
> The difference in expectation is that I want it to diagnose the use of an uninitialized variable in _reachable_ code.

Ah, that's fair. Thanks!
